### PR TITLE
Restore the SDK sample to working order.

### PR DESCRIPTION
**Pull request by lgolding**

### CODE REVIEW

**Comment by lgolding** [comment]
@michaelcfanning is added to the review. #Closed

**Comment by lgolding** [comment]
@EasyRhinoMSFT is added to the review. #Closed

**Comment by lgolding** [comment]
@Evmaus-MS is added to the review. #Closed

**Comment by lgolding** [comment]
@cfaucon is added to the review. #Closed

File: src/Samples/Sarif.Sdk.Sample/Program.cs [hunk 1]
```diff
@@ -41,6 +41,8 @@ static int LoadSarifLogFile(LoadOptions options)
             string logText = File.ReadAllText(options.InputFilePath);
             SarifLog log = JsonConvert.DeserializeObject<SarifLog>(logText, settings);
 
+            Console.WriteLine($"The log file \"{options.InputFilePath}\" contains {log.Runs[0]?.Results.Count} results.");
```
Comment (lgolding) [inline src/Samples/Sarif.Sdk.Sample/Program.cs]: Just so this verb produces some output.

File: src/Samples/Sarif.Sdk.Sample/Program.cs [hunk 2]
```diff
@@ -52,9 +54,9 @@ static int LoadSarifLogFile(LoadOptions options)
         static int CreateSarifLogFile(CreateOptions options)
         {
             // We'll use this source file for several defect results -- the
-			// SampleSourceFiles folder should be at the same level as the project folder
+            // SampleSourceFiles folder should be at the same level as the project folder
```
Comment (lgolding) [inline src/Samples/Sarif.Sdk.Sample/Program.cs]: Tabs to spaces.

File: src/Samples/Sarif.Sdk.Sample/Program.cs [hunk 3]
```diff
@@ -226,7 +239,7 @@ static int CreateSarifLogFile(CreateOptions options)
                         var result = new Result()
                         {
                             RuleId = rule.Id,
-                            AnalysisTarget = new FileLocation { Uri = new Uri(@"file://d:/src/module/foo.dll") }, // This is the file that was analyzed
```
Comment (lgolding) [inline src/Samples/Sarif.Sdk.Sample/Program.cs]: Not MS compliant

File: src/Samples/Sarif.Sdk.Sample/Sarif.Sdk.Sample.csproj [hunk 4]
```diff
@@ -38,7 +38,13 @@
     <Reference Include="Newtonsoft.Json, Version=11.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
       <HintPath>..\packages\Newtonsoft.Json.11.0.2\lib\net45\Newtonsoft.Json.dll</HintPath>
     </Reference>
+    <Reference Include="Sarif, Version=2.0.0.0, Culture=neutral, PublicKeyToken=21a5e83f6f5bb844, processorArchitecture=MSIL">
+      <HintPath>..\packages\Sarif.Sdk.2.0.0-csd.1\lib\net461\Sarif.dll</HintPath>
+    </Reference>
```
Comment (lgolding) [inline src/Samples/Sarif.Sdk.Sample/Sarif.Sdk.Sample.csproj]: Consume SDK as NuGet package.

File: src/Samples/Sarif.Sdk.Sample/packages.config [hunk 5]
```diff
@@ -2,7 +2,9 @@
 <packages>
   <package id="CommandLineParser" version="2.2.1" targetFramework="net461" />
   <package id="Newtonsoft.Json" version="11.0.2" targetFramework="net461" />
```
Comment (EasyRhin0) [inline src/Samples/Sarif.Sdk.Sample/packages.config]: >Newtonsoft.Json" version="11.0.2"  [](start = 15, length = 34)

Larry please move this down to 10.0.3 to match the SDK.

File: src/Samples/Sarif.Sdk.Sample/packages.config [hunk 6]
```diff
@@ -2,7 +2,9 @@
...
+  <package id="Sarif.Sdk" version="2.0.0-csd.1" targetFramework="net461" />
   <package id="System.Collections" version="4.0.11" targetFramework="net461" />
+  <package id="System.Collections.Immutable" version="1.5.0" targetFramework="net461" />
```
Comment (lgolding) [inline src/Samples/Sarif.Sdk.Sample/packages.config]: Dependency of Sarif.Sdk.

File: src/Samples/Sarif.Sdk.Sample/packages.config (hunk 5 above)
Comment (lgolding) [reply src/Samples/Sarif.Sdk.Sample/packages.config]: @EasyRhin0  Thanks, fixed.

File: src/Samples/Sarif.Sdk.Sample/Program.cs (hunk 3 above)
Comment (michaelcfanning) [reply src/Samples/Sarif.Sdk.Sample/Program.cs]: Hooray! My pet peeve addressed.

---
In reply to: [210981653](https://github.com/Microsoft/sarif-sdk/pull/1024#discussion_r210981653) [](ancestors = 210981653)

**Comment by michaelcfanning** [comment]
Please add a build step to our scripts that validates this thing is working. :) Then we won't get into a regressed condition again. After that LGTM

**Comment by lgolding** [comment]
@michaelcfanning Roger wilco on the build step.

**Comment by lgolding** [comment]
@michaelcfanning See changes to BuildAndTest.ps1 to build the sample to avoid regression.

@EasyRhinoMSFT FYI

File: scripts/BuildAndTest.ps1 [hunk 7]
```diff
@@ -14,6 +14,8 @@
     Do not rebuild the SARIF object model from the schema.
 .PARAMETER NoBuild
     Do not build.
+.PARAMETER NoBuildSample
```
Comment (michaelcfanning) [inline scripts/BuildAndTest.ps1]: >NoBuildSample [](start = 11, length = 13)

Why do we need this knob? Is there really a situation where disabling the sample build is of material interest? Who would discover this knob in any case? :)

If you want something like this, you should consider a setting for 'Comprehensive' or 'FullGate' and then have things like this disabled by default for happy path developer use. Those settings make sense 'fast and most relevant pre-check-in' and 'slow and comprehensive, verifies that the full VSTS build can do its thing'.

Comment (lgolding) [reply scripts/BuildAndTest.ps1]: I have a switch like this for every step in BuildAndTest. I agree it's not best. I'll do something like you suggest in a separate PR.

Re: discoverability, it's the same as any PowerShell comment that uses the standard doc comment header:
```
Get-Help .\BuildAndTest.ps1 -Detailed
```

Comment (michaelcfanning) [reply scripts/BuildAndTest.ps1]: Roger.

---
In reply to: [211004442](https://github.com/Microsoft/sarif-sdk/pull/1024#discussion_r211004442) [](ancestors = 211004442)

Comment (lgolding) [reply scripts/BuildAndTest.ps1]: @michaelcfanning Come to think of it, I use these knobs all the time when testing changes to the script. But yes, there's probably a better way.

**Comment by lgolding** [comment]
@michaelcfanning Closing the loop: per your comment "Please add a build step to our scripts that validates this thing is working...", we now build the sample in `BuildAndTest.ps1`, both on the desktop and in AppVeyor.

Do you think it's worth adding tests of the sample's functionality? All it does is
1. `load` verb: Read in a SARIF log file and print out how many results it contains.
2. `create` verb" Serialize a SARIF log file from a hard-coded object model.

IMO it's not worth it. If it builds at all, it's hard to see how it can fail to do what it says.

**Pull request merged**